client/components/gridicon/index.tsx

```diff
@@ -11,11 +10,19 @@ import {
 } from 'gridicons/dist/util/icons-offset';
 import spritePath from 'gridicons/svg-sprite/gridicons.svg';
 
-function needsOffset( name, icons ) {
+function needsOffset( name: string, icons: [string] ) {
 	return icons.indexOf( name ) >= 0;
 }
 
-const Gridicon = React.forwardRef( ( props, ref ) => {
+interface Props {
+	icon: string;
```

Comment:
It would be very nice to list these out (I generated this from http://automattic.github.io/gridicons/gridicons.svg):

```ts
type IconIdentifier =
	| 'add-image'
	| 'add-outline'
	| 'add'
	| 'align-center'
	| 'align-image-center'
	| 'align-image-left'
	| 'align-image-none'
	| 'align-image-right'
	| 'align-justify'
	| 'align-left'
	| 'align-right'
	| 'arrow-down'
	| 'arrow-left'
	| 'arrow-right'
	| 'arrow-up'
	| 'aside'
	| 'attachment'
	| 'audio'
	| 'bell'
	| 'block'
	| 'bold'
	| 'book'
	| 'bookmark-outline'
	| 'bookmark'
	| 'briefcase'
	| 'bug'
	| 'calendar'
	| 'camera'
	| 'caption'
	| 'cart'
	| 'chat'
	| 'checkmark-circle'
	| 'checkmark'
	| 'chevron-down'
	| 'chevron-left'
	| 'chevron-right'
	| 'chevron-up'
	| 'clear-formatting'
	| 'clipboard'
	| 'cloud-download'
	| 'cloud-outline'
	| 'cloud-upload'
	| 'cloud'
	| 'code'
	| 'cog'
	| 'comment'
	| 'computer'
	| 'coupon'
	| 'create'
	| 'credit-card'
	| 'crop'
	| 'cross-circle'
	| 'cross-small'
	| 'cross'
	| 'custom-post-type'
	| 'customize'
	| 'domains'
	| 'dropdown'
	| 'ellipsis-circle'
	| 'ellipsis'
	| 'external'
	| 'filter'
	| 'flag'
	| 'flip-horizontal'
	| 'flip-vertical'
	| 'folder-multiple'
	| 'folder'
	| 'fullscreen-exit'
	| 'fullscreen'
	| 'gift'
	| 'globe'
	| 'grid'
	| 'heading-h1'
	| 'heading-h2'
	| 'heading-h3'
	| 'heading-h4'
	| 'heading-h5'
	| 'heading-h6'
	| 'heading'
	| 'heart-outline'
	| 'heart'
	| 'help-outline'
	| 'help'
	| 'history'
	| 'house'
	| 'image-multiple'
	| 'image-remove'
	| 'image'
	| 'indent-left'
	| 'indent-right'
	| 'info-outline'
	| 'info'
	| 'ink'
	| 'institution'
	| 'italic'
	| 'layout-blocks'
	| 'layout'
	| 'line-graph'
	| 'link-break'
	| 'link'
	| 'list-checkmark'
	| 'list-ordered-rtl'
	| 'list-ordered'
	| 'list-unordered'
	| 'location'
	| 'lock'
	| 'mail'
	| 'mention'
	| 'menu'
	| 'menus'
	| 'microphone'
	| 'minus-small'
	| 'minus'
	| 'money'
	| 'multiple-users'
	| 'my-sites-horizon'
	| 'my-sites'
	| 'nametag'
	| 'next-page'
	| 'not-visible'
	| 'notice-outline'
	| 'notice'
	| 'offline'
	| 'pages'
	| 'pause'
	| 'pencil'
	| 'phone'
	| 'pin'
	| 'plans'
	| 'play'
	| 'plugins'
	| 'plus-small'
	| 'plus'
	| 'popout'
	| 'posts'
	| 'print'
	| 'product-downloadable'
	| 'product-external'
	| 'product-virtual'
	| 'product'
	| 'quote'
	| 'read-more'
	| 'reader-follow-conversation'
	| 'reader-follow'
	| 'reader-following-conversation'
	| 'reader-following'
	| 'reader'
	| 'reblog'
	| 'redo'
	| 'refresh'
	| 'refund'
	| 'reply'
	| 'resize'
	| 'rotate'
	| 'scheduled'
	| 'search'
	| 'share-computer'
	| 'share-ios'
	| 'share'
	| 'shipping'
	| 'shutter'
	| 'sign-out'
	| 'site'
	| 'spam'
	| 'speaker'
	| 'special-character'
	| 'star-outline'
	| 'star'
	| 'stats-alt'
	| 'stats-down-alt'
	| 'stats-down'
	| 'stats-up-alt'
	| 'stats-up'
	| 'stats'
	| 'status'
	| 'strikethrough'
	| 'sync'
	| 'tablet'
	| 'tag'
	| 'text-color'
	| 'themes'
	| 'thumbs-up'
	| 'time'
	| 'trash'
	| 'trophy'
	| 'types'
	| 'underline'
	| 'undo'
	| 'user-add'
	| 'user-circle'
	| 'user'
	| 'video-camera'
	| 'video-remove'
	| 'video'
	| 'visible'
	| 'zoom-in'
	| 'zoom-out';

interface Props extends React.SVGAttributes< SVGSVGElement > {
	icon: IconIdentifier;
}
```

Makes for very helpful suggestions ✨ 

![Screen Shot 2019-06-26 at 17 19 57](https://user-images.githubusercontent.com/841763/60192805-cf97d180-9836-11e9-9eea-85aa39551fd5.png)

I am a bit hesitant to maintain this list separate from the source (https://github.com/Automattic/gridicons). Did you think about adding types to the gridicons package? I don't think the icon set changes _too_ frequently, so it may not be an issue.

Reply:
I think there are two separate questions here: whether adding types to the `gridicons` package is a good idea, and whether the list of available icons should be known at compile time.

For the former, I'm generally in support. I'd keep the svg file definitions local to Calypso (since that's dependant on the build system), but the offset stuff can definitely move to `gridicons`.

However, I'm not so sure about the answer to the second question. It *may* make sense to list the icons statically, if we're able to automate that as part of the `gridicons` (and others') build process? Maybe?

Reply:
> and whether the list of available icons should be known at compile time.

Which compile time? Gridicons or Calypso? For a consumer like Calypso, it seems very valuable to know which icons are available and whether an invalid icon has been used.

> It may make sense to list the icons statically, if we're able to automate that as part of the gridicons (and others') build process? Maybe?

I'm not familiar with the gridicons source, but if we use TypeScript strategically there, we can have `tsc` generate definitions.